### PR TITLE
#972: Refactored exception handling for Script Options parser

### PR DESCRIPTION
**Pull request by tomuben**
related to https://github.com/exasol/script-languages-release/issues/972

### CODE REVIEW

File: exaudfclient/base/javacontainer/javacontainer_impl.cc [hunk 1]
```diff
@@ -35,18 +35,21 @@ JavaVMImpl::JavaVMImpl(bool checkOnly, bool noJNI, SwigFactory& swigFactory)
...
 
-    for (set<string>::iterator it = extractor.getJarPaths().begin(); it != extractor.getJarPaths().end();
-         ++it) {
-        addJarToClasspath(*it);
+        for (set<string>::iterator it = extractor.getJarPaths().begin(); it != extractor.getJarPaths().end();
+             ++it) {
+            addJarToClasspath(*it);
+        }
+    } catch(const std::runtime_error& ex) {
+        throwException(ex);
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/javacontainer_impl.cc:52]: I don't think, we actually need to call throwException here, it was more a shortcut, when we init the Jvm and need to throw an exception after a post condition wasn't fulfilled. Normal Exceptions will be caught by JavaVM and then transmitted over the linker namespace.

File: exaudfclient/base/javacontainer/script_options/parser_legacy.cc [hunk 2]
```diff
@@ -126,66 +131,64 @@ void ScriptOptionLinesParserLegacy::extractImportScripts(SwigFactory & swigFacto
...
-                                 std::function<void(const std::string&)> throwException) {
-   parseForMultipleOptions(m_keywords.jvmOptionKeyword(),
-                            [&](const std::string& value, size_t pos){callback(value);},
-                            [&](const std::string& msg){throwException("F-UDF-CL-SL-JAVA-1612" + msg);});
+void ScriptOptionLinesParserLegacy::parseForJvmOptions(std::function<void(const std::string &option)> callback) {
+    try {
+       parseForMultipleOptions(m_keywords.jvmOptionKeyword(),
+                                [&](const std::string& value, size_t pos){callback(value);});
+    } catch(const std::runtime_error& ex) {
+        throw std::runtime_error(std::string("F-UDF-CL-SL-JAVA-1612") + ex.what());
```
Comment (tkilias) [inline exaudfclient/base/javacontainer/script_options/parser_legacy.cc:148]: Maybe a rethrow function might make sense

Comment (tkilias) [reply exaudfclient/base/javacontainer/script_options/parser_legacy.cc:148]: Or crazy function that gets the code in the try block as lambda, returns the return value and catches exceptions and rethrows them with a additional error code

Comment (tkilias) [reply exaudfclient/base/javacontainer/script_options/parser_legacy.cc:148]: Also thinking about if we should through a own exception like a OptionParserException

File: exaudfclient/base/javacontainer/javacontainer_impl.cc (hunk 1 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/javacontainer_impl.cc:52]: Then I can remove the try/catch here

File: exaudfclient/base/javacontainer/script_options/parser_legacy.cc (hunk 2 above)
Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/parser_legacy.cc]: Added a `rethrow` method.
-> The lambda maybe makes the code more difficult to read.

Comment (tomuben) [reply exaudfclient/base/javacontainer/script_options/parser_legacy.cc]: Created class OptionParserException

**Review by tkilias — APPROVED** [review]

**Pull request merged**